### PR TITLE
[Refactor] Modulize sharding methods

### LINKED ISSUE #10: [Bug] Multiple reductions of linear bias after row sharding the weight
Consider the following case, we have partial inputs and shard the weight $A$ of the linear layer by row.

$$
XA+b
=\begin{bmatrix}X_1 & X_2\end{bmatrix}
\begin{bmatrix}A_1 \\\\ A_2\end{bmatrix}
+b
= X_1A_1+X_2A_2+b
$$

In the current implementation (as shown below), we directly conduct `all_reduce` for the output of the linear layer. However, since the bias $b$ is implicitly included in the linear layer, this leads to redundant additions for the bias and generates incorrect results.

```python
class Model(nn.Module):
  def __init__(self):
    self.linear = nn.linear(...)
  def forward(self, x):
    x = self.linear(x)
    dist.all_reduce(x, op=dist.ReduceOp.SUM)
    return x
```

As a result, we need to preprocess the bias when the corresponding linear layer is row sharded. There are two possible solutions:
1. Only one device holds the bias data and other devices set bias to None, so the bias will only be added once.
2. Average the bias across different devices. In that case, the summation of the new bias will be the same as the original one.

### PR DESCRIPTION
**Pull request by comaniac**
## Description ##

Close #10 

This PR refactors shard ops by modulizing necessary methods. In this way, we could easily customize the behavior of sharding a certain op/module. Accordingly, we also ported PR #23 as a demonstration.

Detail changes:
1. A new op `LinearWithSyncFunc`.
2. An API "register_shard_method" that allows each module to register a set of sharding related methods.
3. Two APIs "apply_shard_method" and "apply_sync_method" dispatches to the right method based on the module type.
4. Customize the sharding of nn.Linear and Conv1D.
5. To fix #10, in addition to performing sync op before bias addition, we also have to perform an additional allreduce on the bias gradient. Now this is achieved by annotating the bias parameters, just like layer norm.

## Checklist ##

- [x] PR's title starts with a category (e.g. [Bugfix], [Model], [Tutorial], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

cc @szhengac @chhzh123

### CODE REVIEW

File: slapo/primitives/tensor_parallel.py [hunk 1]
```diff
@@ -96,75 +70,16 @@ def _shard(name, tensor):
...
-                except KeyError:
-                    raise RuntimeError(
-                        f"Output of {sch.path} has to be gathered along axis "
-                        f"{sch.metadata.primitives['shard']['gather_axis']}, but "
-                        f"{gather_axis} is requested"
-                    ) from None
-
-        out_type, out_part_axis = postproc_sharding(
-            sch.mod, tensor_name, sharded_size, axis
+        apply_shard_method("postproc", sch, tensor_name, sharded_size, axis)
```
Comment (szhengac) [inline slapo/primitives/tensor_parallel.py:73]: I feel like the new design makes what was easy a bit over complicated. 

1. What is the relationship between apply_shard_method and ShardMethod?
2. What is the api in ShardMethod such as postproc and sync used for? It seems to me these are in the old schedule's shard or sync API? And what is the relationship between this sync and old sync in schedule?

Comment (comaniac) [reply slapo/primitives/tensor_parallel.py:73]: I agree that the new design is not that straightforward, but I believe this reduces the possible ad-hoc changes in the future. For example, the previous PR #23 has to add many ad-hoc changes specifically for nn.Linear into `def shard`.

1. `apply_shard_method` determines which ShardMethod to be used. For example now we have LinearShardMethod, Conv2dShardMethod, etc. Thus, when `apply_shard_method("postproc", sch, ...)` and `sch.mod` is `nn.Linear`, we will dispatch to the sharding method of `nn.Linear`. This design lets the future developments of sharding method not change the schedule primitive directly.
2. The `postproc` is used to update the attributes (e.g., `linear.in_feature`) after sharding. This is renamed from `get_output_type_after_sharding`. The `sync` is migrated from the old sync primitive. By encapsulating this logic into ShardMethod, we are able to override it for a particular module. For example, when syncing `nn.Linear` with its input feature sharded, we have to apply all-reduce before bias addition, so we cannot simply register a hook. Instead, we replace `nn.Linear` with `LinearWithSyncFunc` and directly insert all_reduce in its forward, so that we could perform all-reduce before bias addition.

File: slapo/sharding/shard_ops.py [hunk 2]
```diff
@@ -1,187 +1,611 @@
...
-    module: torch.nn.Linear
-        The linear layer to shard.
-    param_name: str
-        The name of the parameter to shard.
-    sharded_size: int
-        The size of the sharded dimension.
-    axis: int
-        The axis to shard on.
+    @staticmethod
+    def preproc(sch, param_name, sharded_size, axis):
```
Comment (chhzh123) [inline slapo/sharding/shard_ops.py:252]: `preproc` and `postproc` are for forward pass, then how can we handle the backward pass? Did you all reduce the bias gradient as what was done in #23?

Comment (comaniac) [reply slapo/sharding/shard_ops.py:252]: See (5) in description.

**Review by chhzh123 — APPROVED** [review]
LGTM

**Comment by comaniac** [comment]
Thanks @chhzh123
